### PR TITLE
TLDR-417 basics added

### PR DESCRIPTION
**Pull request by NastyBoget**

### CODE REVIEW

File: dedocutils/text_recognition/tesseract_text_recognizer.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
+from typing import Optional
+
+import numpy as np
+import pytesseract
+
+from dedocutils.text_recognition.abstract_text_recognizer import AbstractTextRecognizer
+
+
+class TesseractTextRecognizer(AbstractTextRecognizer):
```
Comment (Travvy88) [inline dedocutils/text_recognition/tesseract_text_recognizer.py:9]: тессеракт же умеет еще детектить ббоксы, да и вообще он умеет делать два шага в одном: детекция + распознавание. Я в hotdog делаю еще абстрактный класс FullOCR как раз для штук, которые два в одном умеют. А в пердоке, например, есть функция apply_tesseract. Мб здесь тоже запилить такую возможность?

File: dedocutils/text_detection/doctr_text_detector/doctr_text_detector.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+import numpy as np
+import torch
+
+from dedocutils.data_structures.bbox import BBox
+from dedocutils.text_detection.abstract_text_detector import AbstractTextDetector
+from dedocutils.text_detection.doctr_text_detector.doctr.models import detection_predictor
+from dedocutils.text_detection.doctr_text_detector.doctr.models.detection.predictor import DetectionPredictor
+
+
+class DoctrTextDetector(AbstractTextDetector):
```
Comment (Travvy88) [inline dedocutils/text_detection/doctr_text_detector/doctr_text_detector.py:13]: здесь еще добавить бы TesseractTextDetector

Comment (NastyBoget) [reply dedocutils/text_detection/doctr_text_detector/doctr_text_detector.py:13]: кмк эту штуку ты можешь добавить отдельным мр если у тебя уже есть код для этого)

File: dedocutils/text_recognition/tesseract_text_recognizer.py (hunk 1 above)
Comment (NastyBoget) [reply dedocutils/text_recognition/tesseract_text_recognizer.py:9]: выглядит логично, сделаю

File: README.md [hunk 3]
```diff
@@ -1,2 +1,33 @@
...
-Useful utilities for automatic document images processing
+
+This library contains useful utilities for automatic document images processing.
+
+## Installation
+
+To install the library use the following command:
+
+```bash
+pip install dedocutils
```
Comment (dronperminov) [inline README.md:10]: а тут разве не должно быть `-`?

File: dedocutils/preprocessing/skew_corrector.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    This class is used for automatic skew correction of the document image.
+    It is useful for small angles skew correction.
+    The projection method is used to determine the rotation angle.
+    """
+    def __init__(self) -> None:
+        self.delta = 1  # step
+        self.limit = 45  # max angle
+
+    def preprocess(self, image: np.ndarray, parameters: Optional[dict] = None) -> np.ndarray:
+        result, _ = self.__auto_rotate(image)
```
Comment (dronperminov) [inline dedocutils/preprocessing/skew_corrector.py:20]: `orientation_angle` больше никогда не сможет быть передан, мб сделаем так? `parameters.get("orientation_angle", 0)`
плюс точно ли имеет смысл делать метод самого вращения приватным и делать столь минималистичную обёртку??

File: dedocutils/preprocessing/skew_corrector.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        rotated = self._rotate_image(image, best_angle)
+        return rotated, best_angle + orientation_angle
+
+    def __determine_score(self, arr: np.ndarray, angle: int) -> (np.ndarray, float):
+        data = self._rotate_image(arr, angle)
+        histogram = np.sum(data, axis=1, dtype=float)
+        score = np.sum((histogram[1:] - histogram[:-1]) ** 2, dtype=float)
+        return score
+
+    def _rotate_image(self, image: np.ndarray, angle: float, color_bound: Tuple[int, int, int] = (255, 255, 255)) -> np.ndarray:
```
Comment (dronperminov) [inline dedocutils/preprocessing/skew_corrector.py:51]: точно ли метод должен быть `protected`?

File: dedocutils/text_detection/doctr_text_detector/doctr/__init__.py
Comment (dronperminov) [inline dedocutils/text_detection/doctr_text_detector/doctr/__init__.py:1]: пустой :)

File: dedocutils/text_detection/doctr_text_detector/doctr/models/__init__.py [hunk 6]
```diff
@@ -0,0 +1,3 @@
+# noqa
```
Comment (dronperminov) [inline dedocutils/text_detection/doctr_text_detector/doctr/models/__init__.py:1]: интересная задумка (не претензия)

File: dedocutils/text_detection/doctr_text_detector/doctr/models/_utils.py [hunk 7]
```diff
@@ -0,0 +1,218 @@
+# noqa
```
Comment (dronperminov) [inline dedocutils/text_detection/doctr_text_detector/doctr/models/_utils.py:1]: ещё одна интересная задумка

File: dedocutils/text_detection/doctr_text_detector/doctr/models/classification/predictor/pytorch.py
Comment (dronperminov) [inline dedocutils/text_detection/doctr_text_detector/doctr/models/classification/predictor/pytorch.py:1]: крайне внезапное назване файла (опять же, не претензия)

File: dedocutils/text_detection/doctr_text_detector/doctr/models/classification/predictor/pytorch.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+class CropOrientationPredictor(nn.Module):
+    """Implements an object able to detect the reading direction of a text box.
+    4 possible orientations: 0, 90, 180, 270 degrees counter clockwise.
+
+    Args:
+        pre_processor: transform inputs for easier batched model inference
+        model: core classification architecture (backbone + classification head)
+    """
+
+    def __init__(
```
Comment (dronperminov) [inline dedocutils/text_detection/doctr_text_detector/doctr/models/classification/predictor/pytorch.py:21]: я бы мог, конечно, написать, что тут влезает на одну строку, но не буду, нафиг нафиг пытаться наводить красоту в этом чужом страшилище

File: dedocutils/text_recognition/abstract_text_recognizer.py [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+
+
+class AbstractTextRecognizer(ABC):
+    """
+    Abstract class for OCR on document images.
+    """
+    @abstractmethod
+    def recognize(self, image: np.ndarray, parameters: Optional[dict] = None) -> str:
+        """
+
```
Comment (dronperminov) [inline dedocutils/text_recognition/abstract_text_recognizer.py:14]: лишняя путсая строка

File: tests/__init__.py
Comment (dronperminov) [inline tests/__init__.py:1]: пустой

File: dedocutils/text_recognition/tesseract_text_recognizer.py (hunk 1 above)
Comment (NastyBoget) [reply dedocutils/text_recognition/tesseract_text_recognizer.py:9]: сделала модуль detection_recognition для этого

File: tests/unit_tests/__init__.py
Comment (dronperminov) [inline tests/unit_tests/__init__.py:1]: пустой

File: tests/unit_tests/test_classes.py [hunk 10]
```diff
@@ -0,0 +1,14 @@
+import os.path
+import unittest
+
+import cv2
+from dedocutils.text_detection import DoctrTextDetector
+
+
+class TestClasses(unittest.TestCase):
+
```
Comment (dronperminov) [inline tests/unit_tests/test_classes.py:9]: лишняя пустая строка :)

File: README.md (hunk 3 above)
Comment (NastyBoget) [reply README.md:10]: должно быть, сделаю

File: dedocutils/text_detection/doctr_text_detector/doctr_text_detector.py (hunk 2 above)
Comment (Travvy88) [reply dedocutils/text_detection/doctr_text_detector/doctr_text_detector.py:13]: давай)

File: dedocutils/preprocessing/skew_corrector.py (hunk 5 above)
Comment (NastyBoget) [reply dedocutils/preprocessing/skew_corrector.py:51]: это так было

File: dedocutils/text_detection/doctr_text_detector/doctr/__init__.py
Comment (NastyBoget) [reply dedocutils/text_detection/doctr_text_detector/doctr/__init__.py:1]: я буду придерживаться инитов везде в этом проекте

File: dedocutils/text_detection/doctr_text_detector/doctr/models/__init__.py (hunk 6 above)
Comment (NastyBoget) [reply dedocutils/text_detection/doctr_text_detector/doctr/models/__init__.py:1]: в дедке тоже имеется

File: dedocutils/text_detection/doctr_text_detector/doctr/models/_utils.py (hunk 7 above)
Comment (NastyBoget) [reply dedocutils/text_detection/doctr_text_detector/doctr/models/_utils.py:1]: в дедке тоже имеется

File: dedocutils/preprocessing/skew_corrector.py [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+
+
+class SkewCorrector(AbstractPreprocessor):
+    """
+    This class is used for automatic skew correction of the document image.
+    It is useful for small angles skew correction.
+    The projection method is used to determine the rotation angle.
+    """
+    def __init__(self) -> None:
+        self.delta = 1  # step
```
[the remark this answers is not in the document]
Comment (dronperminov) [reply dedocutils/preprocessing/skew_corrector.py:16]: во-первых, два
я не про параметр в конструктор, а замену `delta -> step` и `limit -> max_angle`

File: dedocutils/text_detection/doctr_text_detector/doctr/models/classification/predictor/pytorch.py (hunk 8 above)
Comment (NastyBoget) [reply dedocutils/text_detection/doctr_text_detector/doctr/models/classification/predictor/pytorch.py:21]: я даже не пыталась (честно)

File: dedocutils/text_recognition/abstract_text_recognizer.py (hunk 9 above)
Comment (NastyBoget) [reply dedocutils/text_recognition/abstract_text_recognizer.py:15]: теперь не пустая

File: tests/unit_tests/__init__.py
Comment (NastyBoget) [reply tests/unit_tests/__init__.py:1]: я буду придерживаться инитов везде в этом проекте

File: tests/unit_tests/test_classes.py (hunk 10 above)
Comment (NastyBoget) [reply tests/unit_tests/test_classes.py:9]: уфффф захотелось отделить название от кода

File: dedocutils/preprocessing/skew_corrector.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (NastyBoget) [reply dedocutils/preprocessing/skew_corrector.py]: сделала

File: dedocutils/preprocessing/skew_corrector.py (hunk 5 above)
Comment (NastyBoget) [reply dedocutils/preprocessing/skew_corrector.py]: сделала `private`

File: tests/unit_tests/__init__.py
Comment (NastyBoget) [reply tests/unit_tests/__init__.py:1]: убрала

File: dedocutils/preprocessing/skew_corrector.py (hunk 4 above)
Comment (dronperminov) [reply dedocutils/preprocessing/skew_corrector.py]: Так она же не рекурсивная вроде...

Comment (NastyBoget) [reply dedocutils/preprocessing/skew_corrector.py]: Видимо я спутала со старой версией, убрала обертку

**Review by dronperminov — APPROVED** [review]
